File: android/runtime/src/main/java/com/amitia/runtime/extension/security/EntryValidator.kt
package com.amitia.runtime.extension.security

class EntryValidator(private val policy: ArchivePolicy) {

    private val forbiddenExtensions = setOf(
        ".exe", ".com", ".bat", ".cmd",
        ".msi", ".dll", ".sys", ".scr",
        ".pif", ".cpl", ".so", ".dylib",
        ".app", ".class", ".jar",
        ".apk", ".deb", ".rpm"
    )

    private val nestedArchiveExtensions = setOf(
        ".zip", ".rar", ".7z", ".tar",
        ".gz", ".tgz", ".bz2", ".xz"
    )

    private val executableMagicBytes = listOf(
        byteArrayOf('M'.code.toByte(), 'Z'.code.toByte()),
        byteArrayOf(0x7f, 'E'.code.toByte(), 'L'.code.toByte(), 'F'.code.toByte()),
        byteArrayOf(0xca.toByte(), 0xfe.toByte(), 0xba.toByte(), 0xbe.toByte()),
        byteArrayOf(0xcf.toByte(), 0xfa.toByte(), 0xed.toByte(), 0xfe.toByte()),
        byteArrayOf(0xce.toByte(), 0xfa.toByte(), 0xed.toByte(), 0xfe.toByte())
    )

    private val archiveMagicBytes = listOf(
        byteArrayOf('P'.code.toByte(), 'K'.code.toByte(), 0x03, 0x04),
        byteArrayOf('P'.code.toByte(), 'K'.code.toByte(), 0x05, 0x06),
        byteArrayOf(0x1f, 0x8b.toByte()),
        byteArrayOf('B'.code.toByte(), 'Z'.code.toByte(), 'h'.code.toByte()),
        byteArrayOf(0xfd.toByte(), '7'.code.toByte(), 'z'.code.toByte(), 'X'.code.toByte(), 'Z'.code.toByte(), 0x00),
        byteArrayOf('R'.code.toByte(), 'a'.code.toByte(), 'r'.code.toByte(), '!'.code.toByte(), 0x1a, 0x07)
    )

    private val secretIndicators = listOf(
        "-----BEGIN RSA PRIVATE KEY-----",
        "-----BEGIN EC PRIVATE KEY-----",
        "-----BEGIN PRIVATE KEY-----",
        "-----BEGIN CERTIFICATE-----",
        "ghp_",
        "gho_",
        "ghu_",
        "ghs_",
        "ghr_",
        "sk-"
    )

    data class EntryValidationResult(
        val passed: Boolean,
        val warnings: List<String> = emptyList(),
        val errors: List<String> = emptyList()
    )

    fun validate(path: String, content: ByteArray): EntryValidationResult {
        val errors = mutableListOf<String>()
        val warnings = mutableListOf<String>()

        val ext = getFileExtension(path).lowercase()

        if (forbiddenExtensions.contains(ext)) {
            errors.add("forbidden file extension: $ext")
        }

        if (nestedArchiveExtensions.contains(ext) && !policy.allowNestedArchive) {
            errors.add("nested archive not allowed: $ext")
        }

        if (ext == ".wasm" && content.size >= 4 &&
            content[0] == 0x00.toByte() && content[1] == 'a'.code.toByte() &&
            content[2] == 's'.code.toByte() && content[3] == 'm'.code.toByte()
        ) {
            warnings.add("WASM binary detected: $path")
        }

        for (magic in executableMagicBytes) {
            if (content.size >= magic.size && content.copyOfRange(0, magic.size).contentEquals(magic)) {
                if (!policy.allowExecutable) {
                    errors.add("executable binary not allowed: $path")
                } else {
                    warnings.add("executable binary: $path")
                }
                break
            }
        }

        for (magic in archiveMagicBytes) {
            if (content.size >= magic.size && content.copyOfRange(0, magic.size).contentEquals(magic)) {
                if (!policy.allowNestedArchive && ext != ".zip" && ext != ".amitiax") {
                    warnings.add("archive magic without expected extension: $path")
                }
                break
            }
        }

        checkSecretPatterns(path, content, warnings)

        return EntryValidationResult(
            passed = errors.isEmpty(),
            warnings = warnings,
            errors = errors
        )
    }

    private fun checkSecretPatterns(path: String, content: ByteArray, warnings: MutableList<String>) {
        if (content.isEmpty()) return

        val text = try {
            String(content, Charsets.UTF_8)
        } catch (e: Exception) {
            return
        }

        for (indicator in secretIndicators) {
            if (text.contains(indicator)) {
                warnings.add("potential secret detected in: $path ($indicator...)")
                break
            }
        }

        val pathLower = path.lowercase()
        if (pathLower == ".env" || pathLower.endsWith(".env")) {
            warnings.add("environment file detected: $path")
        }
    }

    private fun getFileExtension(path: String): String {
        val lastDot = path.lastIndexOf('.')
        val lastSlash = path.lastIndexOf('/')
        if (lastDot > lastSlash && lastDot >= 0) {
            return path.substring(lastDot)
        }
        return ""
    }
}

File: backend/internal/extension/kernel/package_security/entry_validator.go
package package_security

import (
	"bytes"
	"net/http"
	"path"
	"strings"
	"unicode/utf8"
)

type EntryValidator struct {
	policy              ArchivePolicy
	fileTypeDetector    *FileTypeDetector
	forbiddenExt        map[string]bool
	nestedArchiveExt    map[string]bool
	executableMagics    [][]byte
	declaredExecutables map[string]struct{}
}

func NewEntryValidator(policy ArchivePolicy) *EntryValidator {
	return NewEntryValidatorWithDeclaredExecutables(policy, nil)
}

func NewEntryValidatorWithDeclaredExecutables(policy ArchivePolicy, declared map[string]struct{}) *EntryValidator {
	copied := make(map[string]struct{}, len(declared))
	for name := range declared {
		copied[strings.ToLower(strings.ReplaceAll(name, "\\", "/"))] = struct{}{}
	}
	return &EntryValidator{
		policy:           policy,
		fileTypeDetector: NewFileTypeDetector(),
		forbiddenExt: map[string]bool{
			".exe": true, ".com": true, ".bat": true, ".cmd": true,
			".msi": true, ".dll": true, ".sys": true, ".scr": true,
			".pif": true, ".cpl": true, ".so": true, ".dylib": true,
			".app": true, ".class": true, ".jar": true,
			".apk": true, ".deb": true, ".rpm": true,
		},
		nestedArchiveExt: map[string]bool{
			".zip": true, ".rar": true, ".7z": true, ".tar": true,
			".gz": true, ".tgz": true, ".bz2": true, ".xz": true,
		},
		executableMagics: [][]byte{
			{'M', 'Z'},
			{0x7f, 'E', 'L', 'F'},
			{0xca, 0xfe, 0xba, 0xbe},
			{0xcf, 0xfa, 0xed, 0xfe},
			{0xce, 0xfa, 0xed, 0xfe},
		},
		declaredExecutables: copied,
	}
}

type EntryValidationResult struct {
	Passed   bool
	Warnings []string
	Errors   []string
}

func (v *EntryValidator) Validate(entry ArchiveEntryInfo, content []byte) EntryValidationResult {
	result := EntryValidationResult{Passed: true}

	ext := strings.ToLower(path.Ext(entry.NormalizedPath))
	_, declaredExecutable := v.declaredExecutables[strings.ToLower(strings.ReplaceAll(entry.NormalizedPath, "\\", "/"))]
	executableAllowed := v.policy.AllowExecutable || (v.policy.AllowDeclaredExecutable && declaredExecutable)

	if v.forbiddenExt[ext] && !executableAllowed {
		result.Passed = false
		result.Errors = append(result.Errors, "forbidden file extension: "+ext)
	}

	if v.nestedArchiveExt[ext] && !v.policy.AllowNestedArchive {
		result.Passed = false
		result.Errors = append(result.Errors, "nested archive not allowed: "+ext)
	}

	if ext == ".wasm" && len(content) >= 4 && bytes.Equal(content[:4], []byte{0, 'a', 's', 'm'}) {
		result.Warnings = append(result.Warnings, "WASM binary detected: "+entry.NormalizedPath)
	}

	for _, magic := range v.executableMagics {
		if len(content) >= len(magic) && bytes.Equal(content[:len(magic)], magic) {
			if !executableAllowed {
				result.Passed = false
				result.Errors = append(result.Errors, "executable binary not allowed: "+entry.NormalizedPath)
			} else {
				result.Warnings = append(result.Warnings, "declared executable binary: "+entry.NormalizedPath)
			}
			break
		}
	}

	mime := http.DetectContentType(content)
	if !strings.HasPrefix(mime, "text/") &&
		!strings.Contains(mime, "json") &&
		!strings.Contains(mime, "xml") &&
		!strings.HasPrefix(mime, "image/") &&
		!strings.HasPrefix(mime, "audio/") &&
		!strings.HasPrefix(mime, "video/") &&
		mime != "application/octet-stream" &&
		len(content) > 0 {
		result.Warnings = append(result.Warnings, "unknown binary type: "+mime+" at "+entry.NormalizedPath)
	}

	if utf8.Valid(content) && len(content) > 0 {
		v.checkSecretPatterns(entry.NormalizedPath, content, &result)
	}

	return result
}

func (v *EntryValidator) checkSecretPatterns(path string, content []byte, result *EntryValidationResult) {
	text := string(content)

	secretIndicators := []string{
		"-----BEGIN RSA PRIVATE KEY-----",
		"-----BEGIN EC PRIVATE KEY-----",
		"-----BEGIN PRIVATE KEY-----",
		"-----BEGIN CERTIFICATE-----",
		"ghp_",
		"gho_",
		"ghu_",
		"ghs_",
		"ghr_",
		"sk-",
	}

	for _, indicator := range secretIndicators {
		if strings.Contains(text, indicator) {
			result.Warnings = append(result.Warnings, "potential secret detected in: "+path+" ("+indicator+"...)")
			break
		}
	}

	pathLower := strings.ToLower(path)
	if pathLower == ".env" || strings.HasSuffix(pathLower, ".env") {
		result.Warnings = append(result.Warnings, "environment file detected: "+path)
	}
}
